### ISSUE TITLE
[BUG] Anvi'o states are not flexible enough

### ISSUE THREAD

**Issue by meren** [merenlab/anvio#1648]
## The need

A user defined minimal state should work well when imported, even if it does not describe the entire visual interface exhaustively.

## The solution

The interactive interface should take in a state file with a grain of salt. If a piece of information exists in it, it should use it. But if something is missing, it shouldn't give an error and use the defaults instead.

# Example

The archive attached contains the following key files:

* `default-orig.json`: A working JSON file exported using `anvi-export-state`.
* `default-mini.json`: A minimal version of the `default-orig.json`.
* `PROFILE.db`: A profile database that contains `default-mini.json` as its default state.

Running this command does not work:

```
anvi-interactive -p PROFILE.db -c CONTIGS.db
```

Because this is how the minimal state looks like, and many of the state variables are not defined in it (to see the difference, please compare it to the `default-orig.json`):

``` json
{
    "version": "3",
    "angle-max": "180",
    "layers": {
        "gc_content": {
            "type": "intensity",
            "color": "#ff0000",
            "color-start": "#e8e8e8"
        }
    },
    "samples-layers": {
        "default": {
            "percent_mapped_reads": {
                "color": "#00ff00"
            }
        }
    }
}
```

Running this command first overwrites the default state with the original one, and works:

```
anvi-import-state -s default-orig.json -p PROFILE.db -n default
anvi-interactive -p PROFILE.db -c CONTIGS.db
```

One can re-import the minimal state this way to test it:

```
anvi-import-state -s default-mini.json -p PROFILE.db -n default
anvi-interactive -p PROFILE.db -c CONTIGS.db
```

## Beneficiaries

Those who wish to generate minimal state files programmatically to mark things in the interactive interface :)

## Archive for tests

[MINITEST.tar.gz](https://github.com/merenlab/anvio/files/5858191/MINITEST.tar.gz)

**Comment by matthewlawrenceklein** [comment]
@meren 

after doing some manual debugging in the `processState()` function in `main.js`, I believe the root issue here is that `processState()` currently calls `buildSamplesTable()`, passing it the 'samples layer order' and 'samples layers' from state, whether or not they exist. Making this call conditional upon the existence of the above state items now allows the interactive interface to continue processing, and it no longer hangs on 'loading'.

This process also brings into question how anvi'o should handle minified states. For example, the `default-mini.json` file above provides a `samples-layers` data point, but not `samples-layers-order`. We can provide a default value for `samples-layers-order` to pass to `buildSamplesTable`, but that wouldn't necessarily have any relationship to the user's 'samples-layers'. 

My next-step question then, are
 
- should anvi'o complain about any particular pieces of state missing from a user-provided minified file
- should anvi'o make assumptions about items missing from a minified state, and sub in some default values
- should anvi'o attempt to programmatically create any missing state items (ie generating a 'samples layer order' if a 'samples layer' is provided without one

**Comment by meren** [comment]
Thanks for looking into this, @matthewlawrenceklein.

> should anvi'o attempt to programmatically create any missing state items (ie generating a 'samples layer order' if a 'samples layer' is provided without one

I think this is the best option. As a general principle, we should always imagine "what would have happened if there was no state", and recover missing things based on the answer to that. If there are cases that are not possible to recover using defaults or simple programmatic workarounds, then we should be able to tell the user that this state doesn't work because it is missing X. Does it make sense? Do you agree?

Thanks,

**Comment by matthewlawrenceklein** [comment]
addressed in PR #1715
